src/solver/goldensectionsearch/mod.rs

```diff
@@ -0,0 +1,139 @@
+// Copyright 2018-2020 argmin developers
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
+// http://opensource.org/licenses/MIT>, at your option. This file may not be
+// copied, modified, or distributed except according to those terms.
+
+//! # References:
+//!
+//! [Wikipedia](https://en.wikipedia.org/wiki/Golden-section_search)
+
+use crate::prelude::*;
+use serde::{Deserialize, Serialize};
+
+const GOLDEN_RATIO: f64 = 1.61803398874989484820;
+const G1: f64 = -1.0 + GOLDEN_RATIO;
+const G2: f64 = 1.0 - G1;
+
+/// Golden-section search
+///
+/// The golden-section search is a technique for finding an extremum (minimum or maximum) of a
+/// function inside a specified interval.
+///
+/// The method operates by successively narrowing the range of values on the specified interval,
+/// which makes it relatively slow, but very robust. The technique derives its name from the fact
+/// that the algorithm maintains the function values for four points whose three interval widths
+/// are in the ratio 2-φ:2φ-3:2-φ where φ is the golden ratio. These ratios are maintained for each
+/// iteration and are maximally efficient.
+///
+/// The `min_bound` and `max_bound` arguments define values that bracket the expected minimum. The
+/// `init_estimate` argument is the initial estimate of the minimum that is required to be larger
+/// than `min_bound` and smaller than `max_bound`.
+///
+/// # References:
+///
+/// [Wikipedia](https://en.wikipedia.org/wiki/Golden-section_search)
+#[derive(Clone, Serialize, Deserialize)]
+pub struct GoldenSectionSearch<O: ArgminOp> {
+    min_bound: f64,
+    max_bound: f64,
+    init_estimate: f64,
+    tolerance: f64,
+
+    x0: f64,
+    x1: f64,
+    x2: f64,
+    x3: f64,
+    f1: O::Output,
+    f2: O::Output,
+}
+
+impl<O> GoldenSectionSearch<O>
+where
+    O: ArgminOp,
+    O::Output: ArgminZero,
+{
+    /// Constructor
+    pub fn new(min_bound: f64, max_bound: f64) -> Self {
+        GoldenSectionSearch {
+            min_bound,
+            max_bound,
+            init_estimate: 0.0,
+            tolerance: 0.01,
+            x0: min_bound,
+            x1: 0.0,
+            x2: 0.0,
+            x3: max_bound,
+            f1: O::Output::zero(),
+            f2: O::Output::zero(),
+        }
+    }
+
+    /// Set tolerance
+    pub fn tolerance(mut self, tol: f64) -> Self {
+        self.tolerance = tol;
+        self
+    }
+}
+
+impl<O> Solver<O> for GoldenSectionSearch<O>
+where
+    O: ArgminOp<Param = f64, Output = f64>,
+{
```

Comment:
It would be interesting to see if it is possible to solve the `f32`/`f64` problem with the `Float` trait of the crate `num_traits`. 

Something like this:

```rust
impl<O, F> Solver<O> for GoldenSectionSearch<O>
where
    O: ArgminOp<Param = F, Output = F>,
    F: Float,
{
    // ...
}
```

Or, maybe even easier (and more related to your initial question in the issue):

```rust
impl<O, F> Solver<O> for GoldenSectionSearch<O>
where
    O: ArgminOp<Param = F, Output = F>,
    F: PartialEq,
{
    // ...
}
```
(You will probably need to add more trait bounds.)

Both potential solutions may however require you to also be more generic in the definition of `GoldenSectionSearch`.

Reply:
I tried this and it went well, once I figured this line

    F: Float + Serialize + DeserializeOwned,

Everything was now in the type choosed by the user. I had one problem though. `ArgminIterData::cost` is defined as a `f64`, so I can't write

    ArgminIterData::new().param(self.x1).cost(self.f1)
    expected `f64`, found type parameter `F`

I would need to cast it and that's exactly what I don't want to do.

Reply:
Thanks for giving this a try! I'm happy to see that it almost works. I guess now is the time to make `ArgminIterData` generic over `f32` and `f64`. I will try to get at this on the weekend, I hope that is not too late for you.

Reply:
Thanks to you and this huge float branch, I was able to test it properly.